### ISSUE TITLE
Copy to clipboard via button instead of [Ctrl]+[C]

### ISSUE THREAD

**Issue by elrido** [PrivateBin/PrivateBin#12]
@rugk:

It is possible to copy the URL automatically to clipboard using  [HTML5](https://stackoverflow.com/questions/26336138/how-can-i-copy-to-clipboard-in-html5-without-using-flash) & [JavaScript](https://github.com/zenorocha/clipboard.js/).

---

@elrido:

Great! Didn't know that.

**Comment by elrido** [comment]
Note in case the site is down, this would be done by using:

    var copyEvent = new ClipboardEvent('copy', { dataType: 'text/plain', data: 'Data to be copied' } );
    document.dispatchEvent(copyEvent);

Source: http://www.htmlgoodies.com/html5/other/working-with-clipboard-apis-in-html5-web-apps.html

**Comment by elrido** [comment]
I looked into it, but found that (for the safety of the user) this can't be run without interaction as I had initially understood.

Currently we already preselect the text and the user may press [Ctrl]+[C] to copy the link. With this interface (which does not work in Chrome and  we could also add a button the user could click to do the same. Hence my question: Is it worth to clutter the interface with yet another button just to provide a feature we already have (not to speak of the additional code and possibly even yet another JS library)?

**Comment by rugk** [comment]
BTW we could also just use this JS https://github.com/zenorocha/clipboard.js/, which makes it very easy to use the clipboard API.

I think this feature is a nice one and we really do not have a lot of buttons at the URL screen so no one can speak of "cluttering". I think a button on the right of the URL would be really nice.

**Comment by elrido** [comment]
That library has the same issue. Its just a nice wrapper and not really needed anyway and would be yet another JS to load for something that can be done in two lines.

The problem is that it does not work as you think it does: You have to trigger that event on a user action like a mouse click or a button press. So this "feature" adds nothing new: The user already can just press a button (on his/her keyboard or touchscreen) to copy the URL. Adding this code does not make that easier, but instead would confuse the user because there are now two ways of copying the URL press [Ctrl]+[C] or the button. Which one should be used? That IMO is bad design.

**Comment by rugk** [comment]
As for the lib: Okay, if it is really so easy we do not need to use it.

It's something new, because it is more convenient to click on a button than to press `Ctrl+C`.
Laos for keyboards this are of course two keys.
I also have the feeling that not many users notice that the  URL is already selected and they can just press Ctrl+C.
I mean yes: It is written there.
But... here we already have the point from teh UX perspective. Users actually *have to read* through a explanatory text to know how to use it. If you think they do not have to read this and it is intuitive enough, then I may ask you: Why did you include the text?

So on the other hand take the button thing: Every user recognized "Copy" buttons if they are designed in the way all other copy buttons are (you know what icon I am specking of...), so the button is very intuitive. One should only need one second (or maybe less) to get out the purpose of that button and I am sure it is clicked very fast.
Also note: For some [local users](https://en.wikipedia.org/wiki/Luser)* which are very unfamilar with keyboard shortcuts, `Ctrl+C` may even be too hard, but all of them can at least click on a button.

* German: DAU :wink:

**Comment by elrido** [comment]
Actually in german it gets translated to [Strg]+[c] which is odd for the german speaking Swiss who have [Ctrl] instead of [Strg].

**Comment by rugk** [comment]
> odd for the german speaking Swiss who have [Ctrl] instead of [Strg].

Really? Did not knew this.
Interesting fact, but I don't get how this is related to this issue. (It may rather be a new issue: "Make a de-CH translaion". Do you want to open one?)

**Comment by rugk** [comment]
BTW in contrast to what the issue title indicates I am not for removing the selection thing. Of course the URL can still be selected, so users can easily press Ctrl+C. The button should just be an additional feature.

**Comment by elrido** [comment]
> "Make a de-CH translation"

Currently we only provide i18n (_internationalization_ aka translation to main language family e.g. en, de) not l10n (_localization_ which would include formatting changes of numbers in the sizeToHumanReadable() and timeToHumanReadable() functions based on locales like en_GB, de_CH, etc.)

But back to topic: I still don't see the worth of this feature, but may get back to this when there is nothing else to do. Personally I will rather focus on the paste manager feature for an upcoming release (and any bugs of course).

Anybody who wants to implement it please feel free to do so. The change would need to be added [somewhere around here](https://github.com/PrivateBin/PrivateBin/blob/cd02c6e916b9515b5aaf288ed4b4f41fe2084aa7/js/privatebin.js#L1082) or possibly in the [helper.selectText()](https://github.com/PrivateBin/PrivateBin/blob/cd02c6e916b9515b5aaf288ed4b4f41fe2084aa7/js/privatebin.js#L133) method. If you insert a button also all the translations in `i18n/` need to be updated to include the new HTML markup.

**Comment by eneerge** [comment]
Here's something that will work on all browsers since 2015.
Props to: https://stackoverflow.com/questions/400212/how-do-i-copy-to-the-clipboard-in-javascript
```
var copyButton = document.querySelector('#copyhint'); // current id of the ctrl+c text in PrivateBin
copyButton.addEventListener(
  'click'
  ,function(event) {
    var copyTextarea = document.querySelector('#pasteurl'); // currently this is the id of the <a href>
    copyTextarea.focus();
    copyTextarea.select(); // this only works on a text field (fails atm unless you convert the #pasteurl to something like a textarea)
    try {
      var successful = document.execCommand('copy');
      var msg = successful ? 'successful' : 'unsuccessful';
      console.log('Copying text command was ' + msg);
    } catch (err) {
      console.log('Oops, unable to copy');
    }
  }
);

```

**Comment by Haocen** [comment]
Reminder: @Haocen this is recently implemented, contribute the code and make a PR.

**Comment by narFnarF** [comment]
As a user, it's uncommon for text to be already selected and only having to press Ctrl+C. Users are more used to selecting the text manually first. So most users will attempt to select the text using the mouse, making it easy to accidentally click the link while doing so. This is problematic for pastebin that are only readable once, as this will delete the text before sending it to its recipient.

To summarize, imagine this problematic scenario:
1. User create a pastebin that is readable only once
2. User accidentally click the link, which open the pastebin to read it
3. User doesn't notice, copies the URL and send it to Friend
4. Friend attempt to open the link, but the data is gone because it was already seen.

To make it easy for users, I would suggest the following changes:
1. Display the URL in big highlighted textbox, separated to any other text, to make it easy to see in the page
2. Add a "copy URL" button next to the URL box. Clicking it copies the URL to the clipboard. Display a little popup or animated text that says "copied!" for confirmation.
3. For users who already know that the URL can be copied by Ctrl+C, we keep the existing functionality: The URL is already pre-selected in its text box and the user can immediately press Ctrl+C to copy it (without having to use the mouse)

In other words, 2 ways of copying the URL: a button or using Ctrl+C

Don't hesitate to contact me for any UX questions (I'm a designer) 😃

**Comment by rugk** [comment]
I agree, I guess just some implementation is missing here. Maybe some bootstrap components could be used for that, but possibly that would require a bootstrap update first (ref #728).
